include/dr/mhp/containers/rows.hpp

```diff
@@ -0,0 +1,238 @@
+        abs_ind <= dm_->local_rows_indices_.second) { // regular rows
+      return dm_->dm_rows_[index_ + n];
+    }
+    if (abs_ind >= (difference_type)(dm_->local_rows_indices_.first -
+                                     dm_->halo_bounds().prev) &&
+        abs_ind < dm_->local_rows_indices_.first) { // halo prev
+      return dm_->dm_halo_p_rows_[dm_->halo_bounds().prev -
+                                 dm_->local_rows_indices_.first + abs_ind];
+    }
+    if (abs_ind > dm_->local_rows_indices_.second &&
+        abs_ind <= (difference_type)(dm_->local_rows_indices_.second +
+                                     dm_->halo_bounds().next)) { // halo next
+      return dm_->dm_halo_n_rows_[dm_->halo_bounds().next +
+                                 dm_->local_rows_indices_.second - abs_ind];
+    }
+    assert(0);
+  }
+
+  // value_type get() const {
+  //   auto segment_offset = index_ + dm_->halo_bounds_.prev;
+  //   auto value = dm_->win_.template get<value_type>(rank_, segment_offset);
+  //   lib::drlog.debug("get {} =  ({}:{})\n", value, rank_, segment_offset);
+  //   return value;
+  // }
+
+  // void put(const value_type &value) const {
+  //   auto segment_offset = index_ + dm_->halo_bounds_.prev;
+  //   lib::drlog.debug("put ({}:{}) = {}\n", rank_, segment_offset, value);
+  //   dm_->win_.put(value, rank_, segment_offset);
+  // }
+
+  // Comparison
+  bool operator==(const dm_rows_iterator &other) const noexcept {
+    return index_ == other.index_ && dm_ == other.dm_;
+  }
+  auto operator<=>(const dm_rows_iterator &other) const noexcept {
+    return index_ <=> other.index_;
+  }
+
+  // Only these arithmetics manipulate internal state
+  auto &operator-=(difference_type n) {
+    index_ -= n;
+    return *this;
+  }
+  auto &operator+=(difference_type n) {
+    index_ += n;
+    return *this;
+  }
+  difference_type operator-(const dm_rows_iterator &other) const noexcept {
+    return index_ - other.index_;
+  }
+
+  // prefix
+  auto &operator++() {
+    *this += 1;
+    return *this;
+  }
+  auto &operator--() {
+    *this -= 1;
+    return *this;
+  }
+
+  // postfix
+  auto operator++(int) {
+    auto prev = *this;
+    *this += 1;
+    return prev;
+  }
+  auto operator--(int) {
+    auto prev = *this;
+    *this -= 1;
+    return prev;
+  }
+
+  auto operator+(difference_type n) const {
+    auto p = *this;
+    p += n;
+    return p;
+  }
+  auto operator-(difference_type n) const {
+    auto p = *this;
+    p -= n;
+    return p;
+  }
+
+  // When *this is not first in the expression
+  friend auto operator+(difference_type n, const dm_rows_iterator &other) {
+    return other + n;
+  }
+
+  auto segments() const {
+    // return lib::internal::drop_segments(dm_->segments(), index_);
+    return dm_->segments();
+  }
+  auto &halo() const { return dm_->halo(); }
+
+  bool is_local() { return dm_->is_local_row(index_); }
+
+private:
+  DM *dm_ = nullptr;
+  std::size_t index_ = 0;
```

Comment:
`index_` may mean everything, `row_number` is more descriptive IMO

Reply:
row_idx is a compromise, I suppose